Source/WebKit/NetworkProcess/NetworkSession.cpp

```diff
@@ -714,6 +715,59 @@ void NetworkSession::setEmulatedConditions(std::optional<int64_t>&& bytesPerSeco
 
 #endif // ENABLE(INSPECTOR_NETWORK_THROTTLING)
 
+static double connectionTimesMovingAverage(const std::array<Seconds, 25>& connectionTimes, unsigned frontIndex, unsigned entriesCount)
+{
+    constexpr double alphaSmoothing { 0.75 };
+    // EWMA:
+    // s_0 = x_0
+    // s_t = a * x_{t-1} + (1 - a) * s_{t-1}
+    //
+    // Where x_0 is the average of all recent connection timings, and alpha is 0.75.
+
+    double average = std::accumulate(connectionTimes.begin(), connectionTimes.end(), Seconds { }).seconds() / entriesCount;
+    auto front = connectionTimes.begin();
+    std::advance(front, frontIndex);
+    std::for_each(front, connectionTimes.end(), [&average, alphaSmoothing](auto&& timing) {
+        average = alphaSmoothing*timing.seconds() + (1 - alphaSmoothing)*average;
+    });
+    std::for_each(connectionTimes.begin(), front, [&average, alphaSmoothing](auto&& timing) {
+        average = alphaSmoothing*timing.seconds() + (1 - alphaSmoothing)*average;
+    });
+
+    return average;
+}
+
+void NetworkSession::recordHTTPSConnectionTiming(const NetworkLoadMetrics& metrics)
+{
+    constexpr double minumumConnectionTimeout { 3 };
+    constexpr double computedTimeoutScalingFactor { 1.5 };
+
+    if (metrics.reusedConnection())
+        return;
+
+    Seconds connectionEstablishmentTime = metrics.connectEnd - metrics.fetchStart;
```

Comment:
If there is a TLS handshake this will represent 2 roundtrips, but if not it will represent 1 roundtrip.  That will probably mess up the running average.